### PR TITLE
Storage: Avoid wasting CPU with inefficient scheduling in SegmentReadTaskScheduler

### LINKED ISSUE #6495: SegmentReadTaskScheduler waste too many CPU
## Bug Report

image

The reason is that:
- `scheduleSegmentReadTaskPoolUnlock` does not check the limit on the number of active segments. If the number of active segments has reached the limit, this causes `scheduleSegmentUnlock` to fail to schedule a segment.
- When `scheduleSegmentUnlock` scheduling fails, the scheduler will immediately execute the next scheduling because there may be other `segmentReadTaskPool`s to schedule.
- If there are no tasks to schedule, this will cause the scheduling to keep executing between `scheduleSegmentReadTaskPoolUnlock` and `scheduleSegmentUnlock`.

### PR DESCRIPTION
**Pull request by JinheLin**
### What problem does this PR solve?

Issue Number: close #6495 

Problem Summary: 

### What is changed and how it works?
To determine whether a `SegmentReadTaskPool` needs to be scheduled, we need to determine free block slots and free active segments both greater than 0.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
Before this PR:
image
After this PR:
image

- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/DeltaMerge/SegmentReadTaskPool.h [hunk 1]
```diff
@@ -225,7 +228,7 @@ class SegmentReadTaskPool : private boost::noncopyable
 
     int64_t increaseUnorderedInputStreamRefCount();
     int64_t decreaseUnorderedInputStreamRefCount();
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/SegmentReadTaskPool.h:230]: also use `Int64`?

**Review by Lloyd-Pottiger — APPROVED** [review]
Others LGTM.

File: dbms/src/Storages/DeltaMerge/SegmentReadTaskPool.h [hunk 2]
```diff
@@ -183,6 +184,8 @@ class SegmentReadTaskPool : private boost::noncopyable
         , unordered_input_stream_ref_count(0)
         , exception_happened(false)
         , mem_tracker(current_memory_tracker == nullptr ? nullptr : current_memory_tracker->shared_from_this())
+        , block_slot_limit(std::max(num_streams_, 3))
+        , active_segment_limit(std::max(num_streams_, 2))
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/SegmentReadTaskPool.h:188]: Thread number of a SegmentReadPool object is the same as the number of CPU logical core of a CPU NUMA node.
Why `active_segment_limit` and `block_slot_limit` is set to `num_streams_` which mainly equal to half of the number of CPU logical core? When there are two NUMA node, they are the same, but what if there are more than two NUMA node?

Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/SegmentReadTaskPool.h:188]: get

File: dbms/src/Storages/DeltaMerge/SegmentReadTaskPool.h [hunk 3]
```diff
@@ -275,6 +279,9 @@ class SegmentReadTaskPool : private boost::noncopyable
     // std::once_flag and std::call_once to prevent duplicated add.
     std::once_flag add_to_scheduler;
 
+    const Int64 block_slot_limit;
+    const Int64 active_segment_limit;
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/SegmentReadTaskPool.h:283]: Add some comments about these two var?

File: dbms/src/Storages/DeltaMerge/SegmentReadTaskPool.h (hunk 2 above)
Comment (flowbehappy) [inline dbms/src/Storages/DeltaMerge/SegmentReadTaskPool.h:188]: I think we need some comments to explain why `3` and `2` are selected.

**Review by JaySon-Huang — APPROVED** [review]

**Pull request merged**